synedrion/src/uint/public_signed.rs

```diff
@@ -0,0 +1,242 @@
+    /// Note: when multiplying two [`Signed`], the bound on the result is equal to the sum of the bounds of the operands.
+    fn checked_mul(&self, rhs: &Self) -> Option<Self> {
+        let bound = self.bound + rhs.bound;
+        if bound < T::BITS {
+            Some(Self {
+                bound,
+                value: self.value.wrapping_mul(&rhs.value),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Performs the unary - operation.
+    pub fn neg(&self) -> Self {
+        Self {
+            value: T::zero().wrapping_sub(&self.value),
+            bound: self.bound,
+        }
+    }
+}
+
+impl<T> PublicSigned<T>
+where
+    T: Integer + Bounded + Encoding,
+{
+    /// Returns a value in range `[-bound, bound]` derived from an extendable-output hash.
+    ///
+    /// This method should be used for deriving non-interactive challenges,
+    /// since it is guaranteed to produce the same results on 32- and 64-bit platforms.
+    pub fn from_xof_reader_bounded(rng: &mut impl XofReader, bound: &NonZero<T>) -> Self {
+        let bound_bits = bound.as_ref().bits_vartime();
+        assert!(bound_bits < <T as Bounded>::BITS);
+        // Will not overflow because of the assertion above
+        let positive_bound = bound
+            .as_ref()
+            .overflowing_shl_vartime(1)
+            .expect("Just asserted that bound is smaller than precision; qed")
+            .checked_add(&T::one())
+            .expect("does not overflow since we're adding 1 to an even number");
+        let positive_result = uint_from_xof(
+            rng,
+            &NonZero::new(positive_bound).expect("Guaranteed to be greater than zero because we added 1"),
+        );
+        Self::new_from_unsigned(positive_result.wrapping_sub(bound.as_ref()), bound_bits)
+            .expect("Guaranteed to be Some because we checked the bounds just above")
+    }
+}
+
+impl<T> PublicSigned<T>
+where
+    T: Bounded + HasWide + Encoding + Integer,
+    T::Wide: Bounded,
+{
+    /// Returns a [`PublicSigned`] with the same value, but twice the bit-width.
+    pub fn to_wide(&self) -> PublicSigned<T::Wide> {
+        let abs_result = self.abs().to_wide();
+        PublicSigned::new_from_abs(abs_result, self.bound, self.is_negative())
+            .expect("the value fit the bound before, and the bound won't overflow for `WideUint`")
+    }
+}
+
+impl<T> Neg for PublicSigned<T>
+where
+    T: Integer + Bounded,
+{
+    type Output = PublicSigned<T>;
+
+    fn neg(self) -> Self::Output {
+        PublicSigned::neg(&self)
+    }
+}
+
+impl<T> Sub<PublicSigned<T>> for PublicSigned<T>
+where
+    T: Integer + Bounded,
+{
+    type Output = PublicSigned<T>;
+
+    fn sub(self, rhs: PublicSigned<T>) -> Self::Output {
+        self.checked_sub(&rhs)
+            .expect("the calling code ensured the bound is not overflown")
```

Comment:
Hmm, can we do better than assuming the caller knows this? :/

Reply:
The other option is to use `checked_*` everywhere in the proofs/protocols, which will add a lot of code. 

The crucial thing here is that we know that the bounds are static and not data-dependent, so this will panic in tests 100% of the time. But I'm not sure how to articulate it properly.

This is true for `SecretSigned`, but may not be true for `PublicSigned` because they come from external sources. So maybe we can only enable checked ops for these. There's only a few instances where they are used, so perhaps it's the safest course of action.

Reply:
Removed Sub and Mul impls for PublicSigned. They were only used once in `FacProof` creation, so there wasn't a danger of malicious input causing a panic, but we'll be safer this way.